public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx

```diff
@@ -56,25 +63,87 @@ export function MetricSelect({ query, onChange, onGetMetrics }: Props) {
     [styles.highlight]
   );
 
+  const formatLabelFilters = (labelsFilters: QueryBuilderLabelFilter[]): string[] => {
+    return labelsFilters.map((label) => {
+      return `,${label.label}="${label.value}"`;
+    });
+  };
+
+  /**
+   * Transform queryString and any currently set label filters into label_values() string
+   */
+  const queryAndFilterToLabelValuesString = (
+    queryString: string,
+    labelsFilters: QueryBuilderLabelFilter[] | undefined
+  ): string => {
+    return `label_values({__name__=~".*${queryString}"${
+      labelsFilters ? formatLabelFilters(labelsFilters).join() : ''
+    }},__name__)`;
+  };
+
+  /**
+   * There aren't any spaces in the metric names, so let's introduce a wildcard into the regex for each space to better facilitate a fuzzy search
+   */
+  const regexifyLabelValuesQueryString = (query: string) => {
+    const queryArray = query.split(' ');
+    return queryArray.map((query) => `${query}.*`).join('');
+  };
+
+  /**
+   * Reformat the query string and label filters to return all valid results for current query editor state
+   */
+  const formatKeyValueStringsForLabelValuesQuery = (
+    query: string,
+    labelsFilters?: QueryBuilderLabelFilter[]
+  ): string => {
+    const queryString = regexifyLabelValuesQueryString(query);
+
+    return queryAndFilterToLabelValuesString(queryString, labelsFilters);
+  };
+
+  /**
+   * Gets label_values response from prometheus API for current autocomplete query string and any existing labels filters
+   */
+  const getMetricLabels = (query: string) => {
+    // Since some customers can have millions of metrics, whenever the user changes the autocomplete text we want to call the backend and request all metrics that match the current query string
+    const results = datasource.metricFindQuery(formatKeyValueStringsForLabelValuesQuery(query, labelsFilters));
+    return results.then((results) => {
+      if (results.length > MAX_NUMBER_OF_RESULTS) {
+        results.splice(0, results.length - MAX_NUMBER_OF_RESULTS);
+      }
+      return results.map((result) => {
+        return {
+          label: result.text,
```

Comment:
are both label and value required? Can we cut down the size of this object by only passing in one?

Reply:
Nope, both required